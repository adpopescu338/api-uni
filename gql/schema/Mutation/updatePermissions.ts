import { gql } from 'graphql-tag';

export const updatePermissions = gql`
  type Mutation {
    """
    Only sysadmins can update permissions
    """
    updatePermissions(permissions: [UpdatePermissionInput!]!): [Permission!]!
  }

  input UpdatePermissionInput {
    resourceName: String
    operationTypes: [OperationType!]
    id: ID!
  }
`;
