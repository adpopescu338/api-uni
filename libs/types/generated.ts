export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = {
  [_ in K]?: never;
};
export type Incremental<T> =
  | T
  | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string };
  String: { input: string; output: string };
  Boolean: { input: boolean; output: boolean };
  Int: { input: number; output: number };
  Float: { input: number; output: number };
  DateTime: { input: any; output: any };
  _Any: { input: any; output: any };
  _FieldSet: { input: any; output: any };
};

export type ListUsersResponse = {
  __typename?: 'ListUsersResponse';
  nextPage?: Maybe<Scalars['Int']['output']>;
  total: Scalars['Int']['output'];
  users: Array<User>;
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Only sysadmins can create permissions */
  createPermissions: Array<Permission>;
  /** Only sysadmins can create roles */
  createRole: UserDefinedRole;
  /** Only sysadmins can create users */
  createUser: User;
  /** Only sysadmins can delete permissions */
  deletePermission: VoidApiResponse;
  /** Only sysadmins can delete roles */
  deleteRole: VoidApiResponse;
  /**
   * Create a session for a user
   * The session ID is returned as a string
   * This needs to be sent in subsequent requests as a header called 'x-session-id'
   */
  login: Scalars['String']['output'];
  /** Only sysadmins can update permissions */
  updatePermissions: Array<Permission>;
  /** Only sysadmins can update roles */
  updateRole: UserDefinedRole;
  /** If you're not sysadmin, you must be authenticated as the user you're trying to update */
  updateUser: User;
};

export type MutationCreatePermissionsArgs = {
  permissions: Array<PermissionInput>;
};

export type MutationCreateRoleArgs = {
  description?: InputMaybe<Scalars['String']['input']>;
  name: Scalars['String']['input'];
  permissionIds: Array<Scalars['ID']['input']>;
};

export type MutationCreateUserArgs = {
  input: UserInput;
};

export type MutationDeletePermissionArgs = {
  id: Scalars['ID']['input'];
};

export type MutationDeleteRoleArgs = {
  id: Scalars['ID']['input'];
};

export type MutationLoginArgs = {
  email: Scalars['String']['input'];
  password: Scalars['String']['input'];
};

export type MutationUpdatePermissionsArgs = {
  permissions: Array<UpdatePermissionInput>;
};

export type MutationUpdateRoleArgs = {
  description?: InputMaybe<Scalars['String']['input']>;
  id: Scalars['ID']['input'];
  name: Scalars['String']['input'];
  permissionIds: Array<Scalars['ID']['input']>;
};

export type MutationUpdateUserArgs = {
  id: Scalars['ID']['input'];
  input: UserInput;
};

export enum OperationType {
  Create = 'CREATE',
  Delete = 'DELETE',
  Read = 'READ',
  Update = 'UPDATE',
}

export type Permission = {
  __typename?: 'Permission';
  createdAt: Scalars['DateTime']['output'];
  id: Scalars['ID']['output'];
  operationTypes: Array<OperationType>;
  resourceName: Scalars['String']['output'];
  updatedAt?: Maybe<Scalars['DateTime']['output']>;
};

export type PermissionInput = {
  operationTypes: Array<OperationType>;
  resourceName: Scalars['String']['input'];
};

export type Query = {
  __typename?: 'Query';
  _service: _Service;
  /** Only sysadmins can list users */
  listUsers: ListUsersResponse;
  /**
   * Get the current session details (no user details are returned apart from the user ID)
   * This will return an error if the user is not logged in
   */
  session: Session;
  /** Get the current session and user details */
  sessionAndUser: SessionAndUserResponse;
};

export type QueryListUsersArgs = {
  page?: InputMaybe<Scalars['Int']['input']>;
  size?: InputMaybe<Scalars['Int']['input']>;
};

export type Session = {
  __typename?: 'Session';
  createdAt: Scalars['String']['output'];
  expiresAt: Scalars['String']['output'];
  id: Scalars['String']['output'];
  userId: Scalars['String']['output'];
};

export type SessionAndUserResponse = {
  __typename?: 'SessionAndUserResponse';
  session: Session;
  user: User;
};

export type UpdatePermissionInput = {
  id: Scalars['ID']['input'];
  operationTypes?: InputMaybe<Array<OperationType>>;
  resourceName?: InputMaybe<Scalars['String']['input']>;
};

export type User = {
  __typename?: 'User';
  createdAt: Scalars['DateTime']['output'];
  email: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  isSysAdmin: Scalars['Boolean']['output'];
  roles: Array<UserDefinedRole>;
  updatedAt?: Maybe<Scalars['DateTime']['output']>;
};

export type UserDefinedRole = {
  __typename?: 'UserDefinedRole';
  createdAt: Scalars['DateTime']['output'];
  description?: Maybe<Scalars['String']['output']>;
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  permissions: Array<Permission>;
  updatedAt?: Maybe<Scalars['DateTime']['output']>;
};

export type UserInput = {
  email: Scalars['String']['input'];
  isSysAdmin: Scalars['Boolean']['input'];
  name: Scalars['String']['input'];
  password: Scalars['String']['input'];
};

export type VoidApiResponse = {
  __typename?: 'VoidApiResponse';
  success: Scalars['Boolean']['output'];
};

export type _Service = {
  __typename?: '_Service';
  sdl?: Maybe<Scalars['String']['output']>;
};
