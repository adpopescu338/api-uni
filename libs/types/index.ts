export type { Context } from './Context';
export * from './generated';
export { GQLResolver } from './GqlResolver';
export { RedisSession } from './RedisSession';
