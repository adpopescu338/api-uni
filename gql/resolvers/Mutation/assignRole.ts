import { Context } from 'libs/types';
import { MutationAssignRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';

export const assignRole = async (
  _: unknown,
  { userId, roleId }: MutationAssignRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.user.update({
    where: {
      id: userId,
    },
    data: {
      roles: {
        connect: {
          id: roleId,
        },
      },
    },
    include: {
      roles: {
        include: {
          permissions: true,
        },
      },
    },
  });

  return result;
};
