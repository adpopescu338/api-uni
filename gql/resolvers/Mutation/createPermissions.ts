import { Context } from 'libs/types';
import { MutationCreatePermissionsArgs } from 'libs/types/generated';

export const createPermissions = async (
  _: unknown,
  { permissions }: MutationCreatePermissionsArgs,
  { prisma }: Context,
) => {
  const createdPermissions = await prisma.permission.createMany({
    data: permissions,
  });

  return createdPermissions;
};
