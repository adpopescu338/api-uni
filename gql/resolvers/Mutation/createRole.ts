import { Context } from 'libs/types';
import { MutationCreateRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';

export const createRole = async (
  _: unknown,
  { name, description, permissionIds }: MutationCreateRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const role = await prisma.userDefinedRole.create({
    data: {
      name,
      description,
      ...(permissionIds.length && {
        permissions: {
          connect: permissionIds.map((id) => ({ id })),
        },
      }),
    },
  });

  return role;
};
