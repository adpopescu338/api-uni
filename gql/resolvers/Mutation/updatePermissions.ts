import { Permission } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { Context, MutationUpdatePermissionsArgs } from 'libs/types';

export const updatePermissions = async (
  _: unknown,
  { permissions }: MutationUpdatePermissionsArgs,
  { prisma }: Context,
): Promise<Permission[]> => {
  const updates = permissions.map(({ id, operationTypes, resourceName }) => {
    if (!operationTypes?.length && !resourceName) {
      throw new GraphQLError('No update data provided');
    }

    return prisma.permission.update({
      where: {
        id,
      },
      data: {
        ...(operationTypes && {
          operationTypes: {
            set: operationTypes,
          },
        }),
        ...(resourceName && { resourceName }),
      },
    });
  });

  const result = await prisma.$transaction(updates);

  return result;
};
