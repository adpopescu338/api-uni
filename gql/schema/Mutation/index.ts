import { logout } from './logout';
import { login } from './login';

import { createRole } from './createRole';
import { updateRole } from './updateRole';
import { deleteRole } from './deleteRole';
import { assignRole } from './assignRole';

import { createPermissions } from './createPermissions';
import { updatePermissions } from './updatePermissions';
import { deletePermission } from './deletePermission';

import { createUser } from './createUser';
import { updateUser } from './updateUser';

export const Mutation = [
  login,
  logout,

  createRole,
  updateRole,
  deleteRole,
  assignRole,

  createPermissions,
  updatePermissions,
  deletePermission,

  createUser,
  updateUser,
];
