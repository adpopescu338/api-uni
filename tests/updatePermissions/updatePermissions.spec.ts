import { describe, it, expect, jest } from '@jest/globals';
import { updatePermissions } from '../../gql/resolvers/Mutation/updatePermissions';
import { OperationType, UpdatePermissionInput } from 'libs/types/generated';

describe('updatePermissions', () => {
  it('Should be defined', () => {
    expect(updatePermissions).toBeDefined();
  });

  it('Should throw an error if user is not system admin', async () => {
    await expect(
      updatePermissions(undefined, { permissions: [] }, { user: { isSysAdmin: false } } as any),
    ).rejects.toThrowError();
  });

  it('Should save the correct data and return the permissions', async () => {
    const permissions: UpdatePermissionInput[] = [
      {
        id: 'id',
        resourceName: 'someResource',
        operationTypes: [OperationType.Read],
      },
    ];

    const ctx = {
      prisma: {
        permission: {
          update: jest.fn(),
        },
        $transaction: jest.fn().mockResolvedValue(permissions as never),
      },
      user: {
        isSysAdmin: true,
      },
    } as any;

    const result = await updatePermissions(undefined, { permissions }, ctx);
    const { lastCall } = ctx.prisma.permission.update.mock;

    expect(result).toEqual(permissions);
    expect(lastCall).toMatchSnapshot();
  });
});
