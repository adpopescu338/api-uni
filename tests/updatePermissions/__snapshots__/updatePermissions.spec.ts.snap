// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updatePermissions Should save the correct data and return the permissions 1`] = `
[
  {
    "data": {
      "operationTypes": {
        "set": [
          "READ",
        ],
      },
      "resourceName": "someResource",
    },
    "where": {
      "id": "id",
    },
  },
]
`;
